remote.go

```diff
@@ -137,6 +173,99 @@ func (r *Remote) Fetch(o *FetchOptions) (err error) {
 	return r.updateLocalReferenceStorage(o.RefSpecs, refs)
 }
 
+// Push performs a push to the remote. Returns NoErrAlreadyUpToDate if the
+// remote was already up-to-date.
+//
+// TODO: Support deletes.
+// TODO: Support pushing tags.
+// TODO: Check if force update is given, otherwise reject non-fast forward.
+func (r *Remote) Push(o *PushOptions) (err error) {
+	if o.RemoteName == "" {
+		o.RemoteName = r.c.Name
+	}
+
+	if o.RemoteName != r.c.Name {
+		return fmt.Errorf("remote names don't match: %s != %s", o.RemoteName, r.c.Name)
+	}
+
+	if len(o.RefSpecs) == 0 {
+		return fmt.Errorf("no refspec given")
+	}
+
+	if err := o.Validate(); err != nil {
+		return err
+	}
+
+	if r.sendSession == nil {
```

Comment:
We should avoid it entirely, Push and Fetch should be atomic operations without public connect/disconnect methods. However, changing this requires significant changes to how Fetch is used by Repository.